### ISSUE TITLE
[Build] clang15 -Werror -Wdeprecated-builtins fails for abseil-cpp submodule

### ISSUE THREAD

**Issue by jeffdaily** [microsoft/onnxruntime#13676]
### Describe the issue

An upcoming ROCm release will move hip-clang to be based on clang15.  Due to onnxruntime using `-Werror -Wdeprecated-builtins`, any code that uses abseil-cpp type_traits.h and the clang15 compiler will face errors similar to the following:

```
/var/lib/jenkins/onnxruntime/build/RelWithDebInfo/_deps/abseil_cpp-src/absl/meta/type_traits.h:301:36: error: builtin __has_trivial_destructor is deprecated; use __is_trivially_destructible instead [-Werror,-Wdeprecated-builtins]
    : std::integral_constant<bool, __has_trivial_destructor(T) &&
                                   ^
/var/lib/jenkins/onnxruntime/build/RelWithDebInfo/_deps/abseil_cpp-src/absl/meta/type_traits.h:350:36: error: builtin __has_trivial_constructor is deprecated; use __is_trivially_constructible instead [-Werror,-Wdeprecated-builtins]
    : std::integral_constant<bool, __has_trivial_constructor(T) &&
                                   ^
/var/lib/jenkins/onnxruntime/build/RelWithDebInfo/_deps/abseil_cpp-src/absl/meta/type_traits.h:494:17: error: builtin __has_trivial_assign is deprecated; use __is_trivially_assignable instead [-Werror,-Wdeprecated-builtins]
          bool, __has_trivial_assign(typename std::remove_reference<T>::type) &&
                ^
/var/lib/jenkins/onnxruntime/build/RelWithDebInfo/_deps/abseil_cpp-src/absl/meta/type_traits.h:559:8: error: builtin __has_trivial_copy is deprecated; use __is_trivially_constructible instead [-Werror,-Wdeprecated-builtins]
      (__has_trivial_copy(ExtentsRemoved) || !kIsCopyOrMoveConstructible) &&
       ^
/var/lib/jenkins/onnxruntime/build/RelWithDebInfo/_deps/abseil_cpp-src/absl/meta/type_traits.h:560:8: error: builtin __has_trivial_assign is deprecated; use __is_trivially_assignable instead [-Werror,-Wdeprecated-builtins]
      (__has_trivial_assign(ExtentsRemoved) || !kIsCopyOrMoveAssignable) &&
       ^
```

abseil-cpp project seems to have found this issue and corrected it (need to fully verify).

https://github.com/abseil/abseil-cpp/search?q=deprecated-builtins&type=commits

onnxruntime uses an abseil-cpp submodule dated 2021-11-02.  To work with clang15 compiler, either update the abseil-cpp submodule or change the `-Wdeprecated-builtins` to `-Wno-deprecated-builtins`.

### Urgency

We have worked around the issue in our own fork and branch by using `-Wno-deprecated-builtins`.  But this needs to be fixed in general for anyone wishing to use clang15 to compile onnxruntime.

### Target platform

linux, rocm

### Build script

N/A

### Error / output

See above.

### Visual Studio Version

N/A

### GCC / Compiler Version

clang15

**Comment by snnn** [comment]
We don't treat warnings from external libraries as errors, unless something is wrong. For example, on Windows everytime when I compile onnxruntime, I saw tons of C4244 warnings from absl. But they don't fail my build. So, I'm interested to know where the "-Werror" comes from.

**Comment by snnn** [comment]
And does clang15 support CentOS7? Our ROCM pipeline still uses CentOS 7. Should it be changed? @PeixuanZuo I don't understand what role clang plays here.

**Comment by yuslepukhin** [comment]
If the upcoming hip-clang will be based on clang15, then we need to make sure it works, possibly changing pipelines.
Is this blocking you, or will it block in the future. We are not setting `-Werror`.

**Comment by cloudhan** [comment]
I actually met this problem before, when using a homebrew latest hip clang build. Build with `-DCMAKE_CXX_FLAGS=-Wno-deprecated-builtins` will workaround the problem. It is not added to our cmakelists because the `-Wno-deprecated-builtins` is not unknown to old version of hip-clang, causing unknonw flags errors populated.

**Comment by cloudhan** [comment]
@snnn The (hip-)clang is AMD flavored clang, with HIP language support. When building HIP code (cu files in `providers/rocm`), if you include `type_traits.h` from absl, maybe transitively, the compile fails, because `-Werror` is enabled for HIP code. You can add some flags checking and then exclude the warning at https://github.com/microsoft/onnxruntime/blob/b731cf397d5ddc584c6af96af2f14221b93c85a3/cmake/CMakeLists.txt#L1310-L1326

**Comment by snnn** [comment]
I see. "-DCMAKE_CXX_FLAGS=-Wno-deprecated-builtins" is not right. It will override the default CXX flags. For example, if you are doing a release build, you will find "-O2" is missing. Because you define CMAKE_CXX_FLAGS to something, instead of appending something to CMAKE_CXX_FLAGS. 

> either update the abseil-cpp submodule or change the -Wdeprecated-builtins to -Wno-deprecated-builtins.

Yes, we should. Thanks for explanation

**Comment by cloudhan** [comment]
It is passed from commandline, not set in cmakelists.

**Comment by PeixuanZuo** [comment]
Hi @jeffdaily, I found ROCm5.3.x already used clang-15, but when we use this clang-15 under /opt/rocm/llvm/bin/, onnxruntime build well, why we haven't meet this error yet? Is there any difference between the clang-15 being used in ROCm5.3.x and the clang-15 in upcoming ROCm version? Thanks!

/opt/rocm/llvm/bin/clang++ --version
```
AMD clang version 15.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-5.3.0 22362 3cf23f77f8208174a2ee7c616f4be23674d7b081)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm/llvm/bin
```

**Comment by jeffdaily** [comment]
I didn't have any issue using ROCm 5.3.x either, or our soon-to-be-released ROCm 5.4.  Only our current internal development environment (5.5) faces this issue.  In any case, the error is legitimate.  The workaround I used to unblock our dev teams was https://github.com/ROCmSoftwarePlatform/onnxruntime/commit/53c24ad7d3469511b9d742ca97d024c569a854e5.

**Comment by yuslepukhin** [comment]
Updating `abseil-cpp` potentially may cause problems with our CUDA compilation (we have a patch) and natvis updates. It is feasible to do, but may take more time than simply updating the version string.

**Comment by snnn** [comment]
@faxu @pranavsharma , what do you think?

**Comment by pranavsharma** [comment]
Currently we rely on a commit from one of the release branches of abseil. We shouldn't use a commit from the dev branch esp. when there's a workaround for the issue. As @jeffdaily said we can't tell if the dev branch should be considered stable.

**Comment by cloudhan** [comment]
@jeffdaily How is AMD handling the compiling now. In my previous PRs, I move to all HIP code compilation to use native cmake hip language support. So there will be combinations of compiler usage.
1. CMAKE_CXX_COMPILER=gcc and CMAKE_HIP_COMPILER=hipclang, note this is the default if you do not have any config for it
2. CMAKE_CXX_COMPILER=hipclang and CMAKE_HIP_COMPILER=hipclang

If you are still in the first category, then you can try to filter the flag `-Wdeprecated-builtins` for hip translation units in my previous comment (search REMOVE_ITEM in this page). If you fall into the second category, then the only way to go is remove it globaly, maybe condition on when hipclang is used.

**Comment by snnn** [comment]
abseil/googletests are kind of different: https://abseil.io/about/philosophy#upgrade-support

**Comment by snnn** [comment]
> Updating `abseil-cpp` potentially may cause problems with our CUDA compilation (we have a patch) and natvis updates. It is feasible to do, but may take more time than simply updating the version string.

Actually, it will make it better. The patch will be no longer needed. See #13990

**Comment by cloudhan** [comment]
Updating `abseil-cpp` does not fix our problem. The deprecated builtins is still used.

**Comment by snnn** [comment]
While @jeffdaily  said,
> abseil-cpp project seems to have found this issue and corrected it (need to fully verify).

So the answer to this question is: "no, it hasn't". Is it right?

**Comment by TedThemistokleous** [comment]
Are you referring to this fix? https://github.com/microsoft/onnxruntime/pull/13990

**Comment by cloudhan** [comment]
@snnn Yes, abseil actually fixed it (at least some of them, see https://github.com/abseil/abseil-cpp/pull/1289). But we don't have a **release version** that has the fixes been included :(
